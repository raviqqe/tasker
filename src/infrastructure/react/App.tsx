import { PulseLoader } from "react-spinners";
import { useAsync } from "react-use";
import React from "react";
import styled from "styled-components";
import { observer } from "mobx-react";
import { AuthenticationStore } from "../mobx/authentication-store";
import { DocumentsStore } from "../mobx/documents-store";
import { IProps as ILandingProps, Landing } from "./Landing";
import { Home, IProps as IHomeProps } from "./Home";

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  width: 100vw;
`;

interface IProps extends Omit<IHomeProps, "documents">, ILandingProps {
  authenticationStore: AuthenticationStore;
  documentsStore: DocumentsStore;
  initialize: () => Promise<void>;
}

export const App = observer(
  ({
    authenticationStore: { signedIn },
    documentsStore: { documents },
    initialize,
    repositoryURL,
    signIn,
    signOut,
    ...props
  }: IProps) => {
    useAsync(initialize, []);

    return signedIn === null ? (
      <LoaderContainer>
        <PulseLoader color="white" />
      </LoaderContainer>
    ) : signedIn ? (
      <Home {...props} documents={documents} signOut={signOut} />
    ) : (
      <Landing repositoryURL={repositoryURL} signIn={signIn} />
    );
  }
);
