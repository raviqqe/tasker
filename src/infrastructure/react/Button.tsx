import styled from "styled-components";
import { boxShadow } from "./style";
import { red, grey, white } from "./style/colors";

export const Button = styled.button<{ secondary?: boolean }>`
  ${boxShadow};
  background: ${({ secondary }) => (secondary ? grey : red)};
  display: flex;
  justify-content: center;
  align-items: center;
  font-family: inherit;
  color: ${white};
  border: none;
  cursor: pointer;
  flex-shrink: 0;
`;
