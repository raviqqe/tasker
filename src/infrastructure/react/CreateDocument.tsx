import { MdAdd } from "react-icons/md";
import React, { useState } from "react";
import styled from "styled-components";
import { CircleButton } from "./CircleButton";
import { InsertFilesFunction } from "./utilities";
import { MarkdownTextArea } from "./MarkdownTextArea";

const Container = styled.div`
  display: flex;
  align-items: center;
  padding: 1em;
`;

const StyledMarkdownTextArea = styled(MarkdownTextArea)`
  margin-right: -1em;
  max-height: 80vh;
`;

interface IProps {
  createDocument: (text: string) => Promise<void>;
  insertFiles: InsertFilesFunction;
}

export const CreateDocument = ({ createDocument, insertFiles }: IProps) => {
  const [text, setText] = useState("");
  const onSubmit = async (): Promise<void> => {
    setText("");
    await createDocument(text);
  };

  return (
    <Container>
      <StyledMarkdownTextArea
        insertFiles={insertFiles}
        onSubmit={onSubmit}
        setText={setText}
        text={text}
      />
      <CircleButton aria-label="Create" onClick={onSubmit}>
        <MdAdd />
      </CircleButton>
    </Container>
  );
};
