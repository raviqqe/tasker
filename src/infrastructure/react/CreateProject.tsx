import { MdAdd } from "react-icons/md";
import React from "react";
import { CircleButton } from "./CircleButton";

export interface IProps {
  createProject: (name: string) => Promise<void>;
}

export const CreateProject = ({ createProject, ...restProps }: IProps) => (
  <CircleButton
    aria-label="Create Project"
    onClick={async () => {
      const name = window.prompt("Project name?");

      if (!name) {
        return;
      }

      await createProject(name);
    }}
    {...restProps}
  >
    <MdAdd />
  </CircleButton>
);
