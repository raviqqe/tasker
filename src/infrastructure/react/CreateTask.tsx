import { MdAdd } from "react-icons/md";
import React from "react";
import { CircleButton } from "./CircleButton";

interface IProps {
  createTask: (name: string) => Promise<void>;
}

export const CreateTask = ({ createTask }: IProps) => {
  return (
    <CircleButton
      aria-label="Create"
      onClick={async () => {
        const name = window.prompt("Task name?");

        if (!name) {
          return;
        }

        await createTask(name);
      }}
    >
      <MdAdd />
    </CircleButton>
  );
};
