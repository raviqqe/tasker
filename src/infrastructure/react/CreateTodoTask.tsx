import { MdAdd } from "react-icons/md";
import { CircleButton } from "./CircleButton";

export interface IProps {
  createTodoTask: (name: string) => Promise<void>;
}

export const CreateTodoTask = ({ createTodoTask, ...restProps }: IProps): JSX.Element => (
  <CircleButton
    aria-label="Create"
    onClick={async () => {
      const name = window.prompt("Task name?");

      if (!name) {
        return;
      }

      await createTodoTask(name);
    }}
    {...restProps}
  >
    <MdAdd />
  </CircleButton>
);
