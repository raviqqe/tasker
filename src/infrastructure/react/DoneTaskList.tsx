import { styled } from "@linaria/react";
import { defaultImport } from "default-import";
import { useCallback, useEffect, useState } from "react";
import defaultUseInfiniteScroll from "react-infinite-scroll-hook";
import { useAsync, usePrevious } from "react-use";
import type * as domain from "../../domain.js";
import { Loader } from "./Loader.js";
import { Task } from "./Task.js";
import { sleep } from "@raviqqe/loscore/async";

const useInfiniteScroll = defaultImport(defaultUseInfiniteScroll);

const Container = styled.div`
  display: flex;
  flex-direction: column;
  gap: 1rem;
  overflow: auto;
  padding: 1rem;
`;

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`;

export interface Props {
  doneTasks: domain.Task[] | null;
  listMoreDoneTasks: () => Promise<void>;
}

export const DoneTaskList = ({
  doneTasks,
  listMoreDoneTasks,
}: Props): JSX.Element => {
  const [loading, setLoading] = useState(false);
  const [done, setDone] = useState(false);

  const onLoadMore = useCallback(async () => {
    setLoading(true);
    await listMoreDoneTasks();
    await sleep(0);
    setLoading(false);
  }, [listMoreDoneTasks, setLoading]);

  useAsync(onLoadMore, []);

  const [ref] = useInfiniteScroll({
    hasNextPage: !done,
    loading,
    onLoadMore,
  });

  const oldLoading = usePrevious(loading);
  const [length, setLength] = useState(0);

  useEffect(() => {
    if (!oldLoading && loading) {
      setLength(doneTasks?.length ?? 0);
    } else if (oldLoading && !loading && doneTasks?.length === length) {
      setDone(true);
    }
  }, [doneTasks, length, loading, oldLoading]);

  return (
    <Container>
      {(doneTasks ?? []).map((task) => (
        <Task key={task.id} task={task} />
      ))}
      {!done && (
        <LoaderContainer ref={ref}>
          <Loader />
        </LoaderContainer>
      )}
    </Container>
  );
};
