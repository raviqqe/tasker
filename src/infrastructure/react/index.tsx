import { render } from "react-dom";
import React from "react";
import { ApplicationInitializer } from "../../application/application-initializer";
import { DocumentCreator } from "../../application/document-creator";
import { DocumentLister } from "../../application/document-lister";
import { DocumentUpdater } from "../../application/document-updater";
import { IDocument } from "../../domain/document";
import { SignInManager } from "../../application/sign-in-manager";
import { SignOutManager } from "../../application/sign-out-manager";
import { TextFileInserter } from "../../application/text-file-inserter";
import { AuthenticationStore } from "../mobx/authentication-store";
import { DocumentsStore } from "../mobx/documents-store";
import { GlobalStyle } from "./style";
import { App } from "./App";

export class ReactRenderer {
  constructor(
    private readonly applicationInitializer: ApplicationInitializer,
    private readonly documentCreator: DocumentCreator,
    private readonly documentLister: DocumentLister,
    private readonly documentUpdater: DocumentUpdater,
    private readonly signInManager: SignInManager,
    private readonly signOutManager: SignOutManager,
    private readonly textFileInserter: TextFileInserter,
    private readonly authenticationStore: AuthenticationStore,
    private readonly documentsStore: DocumentsStore,
    private readonly repositoryURL: string
  ) {}

  public render(element: HTMLElement): void {
    render(
      <>
        <App
          authenticationStore={this.authenticationStore}
          documentsStore={this.documentsStore}
          createDocument={(text: string) => this.documentCreator.create(text)}
          initialize={() => this.applicationInitializer.initialize()}
          insertFiles={(
            text: string,
            position: number,
            files: File[]
          ): Promise<string> =>
            this.textFileInserter.insert(text, position, files)
          }
          listDocuments={() => this.documentLister.list()}
          listMoreDocuments={() => this.documentLister.listMore()}
          repositoryURL={this.repositoryURL}
          signIn={() => this.signInManager.signIn()}
          signOut={() => this.signOutManager.signOut()}
          updateDocument={(document: IDocument, text: string) =>
            this.documentUpdater.update(document, text)
          }
        />
        <GlobalStyle />
      </>,
      element
    );
  }
}
