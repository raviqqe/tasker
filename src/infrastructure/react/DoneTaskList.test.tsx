import { render } from "@testing-library/react";
import { expect, it } from "vitest";
import { DoneTaskList } from "./DoneTaskList.js";

it("renders", () => {
  expect(
    render(<DoneTaskList doneTasks={[]} listMoreDoneTasks={async () => {}} />)
      .container.firstChild,
  ).toMatchSnapshot();
});
