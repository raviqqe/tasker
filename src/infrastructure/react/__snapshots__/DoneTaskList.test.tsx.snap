// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders 1`] = `
<div
  class="mocked-styled-4 cxbg86i"
>
  <div
    class="mocked-styled-5 l4pu6bf"
  >
    <span
      style="display: inherit;"
    >
      <span
        style="background-color: white; width: 15px; height: 15px; margin: 2px; border-radius: 100%; display: inline-block; animation: react-spinners-PulseLoader-pulse 0.75s 0.12s infinite cubic-bezier(0.2, 0.68, 0.18, 1.08); animation-fill-mode: both;"
      />
      <span
        style="background-color: white; width: 15px; height: 15px; margin: 2px; border-radius: 100%; display: inline-block; animation: react-spinners-PulseLoader-pulse 0.75s 0.24s infinite cubic-bezier(0.2, 0.68, 0.18, 1.08); animation-fill-mode: both;"
      />
      <span
        style="background-color: white; width: 15px; height: 15px; margin: 2px; border-radius: 100%; display: inline-block; animation: react-spinners-PulseLoader-pulse 0.75s 0.36s infinite cubic-bezier(0.2, 0.68, 0.18, 1.08); animation-fill-mode: both;"
      />
    </span>
  </div>
</div>
`;
